### PR TITLE
Exlib malloc branch

### PR DESCRIPTION
**Pull request by PROgram52bc**
Experiment simulating malloc failure.

### CODE REVIEW

**Comment by PROgram52bc** [comment]
And I think `malloc` is a good abstraction level comparing to `brk`, `sbrk`, etc. because coreutils calls `malloc` 706 times, but never calls `brk` or `sbrk` directly.

**Comment by PROgram52bc** [comment]
@Kraks This branch is ready to merge. However the history is kind of ugly.... Maybe I will clean it up sometime.

File: dev-clean/headers/llsc.hpp [hunk 1]
```diff
@@ -553,14 +563,204 @@ class PC {
...
+ * How do we model that rule? Is there a usecase for that?
+ * Use a particular data structure? <2021-10-12, David Deng> */
+
+// An opened file
+struct Stream {
+  private:
+    File file;
+    int mode; // a combination of O_RDONLY, O_WRONLY, O_RDWR, etc.
+    size_t cursor;
+    using status_t = unsigned short;
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:636]: lift `status_t` to top-level?

File: dev-clean/headers/llsc.hpp [hunk 2]
```diff
@@ -553,14 +563,204 @@ class PC {
...
+
+class FS {
+  private:
+    /* TODO: How does immer handle memory leak if map.find returns a raw pointer? <2021-10-12, David Deng> */
+    /* NOTE: Use at instead of find for immer::map <2021-10-12, David Deng> */
+
+    immer::map<Fd, Stream> open_files;
+    immer::map<std::string, File> files;
+    Fd next_fd;
+    Fd open_fd; // set by open()
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:690]: what is `open_fd`?

File: dev-clean/headers/llsc.hpp [hunk 3]
```diff
@@ -627,7 +829,7 @@ inline const BlockLabel mt_bb = 0;
 inline const SS mt_ss = SS(mt_mem, mt_stack, mt_pc, mt_bb);
 
 inline const immer::flex_vector<std::pair<SS, PtrVal>> mt_path_result =
-  immer::flex_vector<std::pair<SS, PtrVal>>{};
+immer::flex_vector<std::pair<SS, PtrVal>>{};
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:832]: indentation

File: dev-clean/headers/llsc.hpp [hunk 4]
```diff
@@ -678,17 +880,17 @@ auto create_async(std::function<T()> f) -> std::future<T> {
...
+      num_async--;
+      lk.unlock();
+      return t;
+      });
   return fu;
 }
 
 // STP interaction
-inline bool use_solver = true;
+inline int use_solver = true;
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:893]: why `use_solver` is changed to `int`?

File: dev-clean/headers/llsc.hpp [hunk 5]
```diff
@@ -992,16 +1194,16 @@ struct CoverageMonitor {
       size_t covered = 0;
       for (auto v : block_cov) { if (v != 0) covered++; }
       std::cout << "#blocks: "
-                << covered << "/"
-                << num_blocks << "; "
-                << std::flush;
+        << covered << "/"
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:1197]: nit: I would prefer the aligned indentation

File: dev-clean/headers/llsc.hpp [hunk 6]
```diff
@@ -1018,46 +1220,85 @@ struct CoverageMonitor {
...
           print_time();
           print_block_cov();
           print_path_cov(false);
           print_async();
           print_query_num();
           std::this_thread::sleep_for(seconds(1));
-        }
-      }).detach();
+          }
+          }).detach();
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:1231]: indentation

File: dev-clean/headers/llsc.hpp [hunk 7]
```diff
@@ -1068,8 +1309,8 @@ sym_exec_br(SS ss, SExpr t_cond, SExpr f_cond,
     if (can_par()) {
       std::future<immer::flex_vector<std::pair<SS, PtrVal>>> tf_res =
         create_async<immer::flex_vector<std::pair<SS, PtrVal>>>([&]{
-          return tf(tbr_ss);
-        });
+            return tf(tbr_ss);
+            });
```
Comment (Kraks) [inline dev-clean/headers/llsc.hpp:1313]: indentation

File: dev-clean/headers/llsc_external.hpp [hunk 8]
```diff
@@ -50,6 +56,36 @@ inline immer::flex_vector<std::pair<SS, PtrVal>> realloc(SS state, immer::flex_v
   return immer::flex_vector<std::pair<SS, PtrVal>>{{res, memLoc}};
 }
 
+inline std::string get_string(PtrVal ptr, SS state) {
+  std::string name;
+  char c = static_cast<char>(proj_IntV(state.at(ptr))); // c = *ptr
```
Comment (Kraks) [inline dev-clean/headers/llsc_external.hpp:61]: truncation happens here -- probably add an assertion that checks the bit-width of `state.at(ptr)`.

File: dev-clean/headers/llsc_external.hpp [hunk 9]
```diff
@@ -50,6 +56,36 @@ inline immer::flex_vector<std::pair<SS, PtrVal>> realloc(SS state, immer::flex_v
...
+
+inline immer::flex_vector<std::pair<SS, PtrVal>> open(SS state, immer::flex_vector<PtrVal> args) {
+  PtrVal ptr = args.at(0);
+  std::string name = get_string(ptr, state);
+  /* TODO: add flags for open_file <2021-11-03, David Deng> */
+  FS fs = state.get_fs().open_file(name, 0);
+  if (fs.status_open_fail()) {
+    return immer::flex_vector<std::pair<SS, PtrVal>>{{state, make_IntV(-1)}};
+  }
+  auto fd = fs.get_open_fd();
```
Comment (Kraks) [inline dev-clean/headers/llsc_external.hpp:78]: what's the value that `fd` is supposed to be?

File: dev-clean/headers/sai.hpp [hunk 10]
```diff
@@ -22,6 +22,8 @@
 
 // #define DEBUG
 
+// #define EXLIB_FAILURE_BRANCH
```
Comment (Kraks) [inline dev-clean/headers/sai.hpp:25]: nit: remove

**Review by Kraks — COMMENTED** [review]
Thanks! Generally looks good -- a few minor comments attached.

Also make sure the code is tested with at least `openfile.c`

**Comment by Kraks** [comment]
To nicify history: `Squash and merge` will combine all commits in this PR into one commit

**Comment by PROgram52bc** [comment]
See #22.

**Pull request closed without merging**